arangod/Metrics/ClusterMetricsFeature.cpp

```diff
@@ -68,73 +73,218 @@ void ClusterMetricsFeature::validateOptions(
-  auto& nf = server().getFeature<NetworkFeature>();
+  // We want more time than _timeout should expire
+  // after last try cross cluster communication
+  _lastUpdate = std::chrono::steady_clock::now();
+  uint64_t version = 0;
+  auto result = readData(version);
+  if (wasStop()) {
+    return;
+  }
+  _lastUpdate = std::chrono::steady_clock::now();
+  if (result == Result::Error) {
+    return rescheduleUpdate(std::max(_timeout, uint32_t{1}));
+  }
+  if (result == Result::New) {
+    return repeatUpdate();
+  }
   auto& cf = server().getFeature<ClusterFeature>();
-  // TODO try get from agency
+  auto& ci = cf.clusterInfo();
+  if (_acquiredUpdateLock) {
+    _acquiredUpdateLock = !ci.unlockMetricsUpdate();
+  }
+  if (_acquiredUpdateLock) {
+    return rescheduleUpdate(std::max(_timeout, uint32_t{1}));
+  }
+  _acquiredUpdateLock = ci.lockMetricsUpdate();
+  if (!_acquiredUpdateLock) {
+    return rescheduleUpdate(std::max(_timeout, uint32_t{1}));
+  }
+  if (wasStop()) {
+    return;
+  }
+  auto& nf = server().getFeature<NetworkFeature>();
   metricsOnCoordinator(nf, cf).thenFinal(
-      [this](futures::Try<RawDBServers>&& metrics) mutable {
-        // We want more time than kTimeout should expire
-        // after last try cross cluster communication
-        _lastUpdate = std::chrono::steady_clock::now();
-        if (!metrics.hasValue()) {
-          scheduleUpdate();
+      [this, version, &ci](futures::Try<RawDBServers>&& raw) mutable {
+        if (wasStop()) {
           return;
         }
+        auto result = Result::Old;
         try {
-          set(std::move(*metrics));
+          _lastUpdate = std::chrono::steady_clock::now();
+          result = writeData(version, std::move(raw));
+          _lastUpdate = std::chrono::steady_clock::now();
         } catch (...) {
+          result = Result::Error;
         }
-        repeatUpdate();
+        _acquiredUpdateLock = !ci.unlockMetricsUpdate();
+        _lastUpdate = std::chrono::steady_clock::now();
+        if (result != Result::Error) {
+          repeatUpdate();
+        }
+        rescheduleUpdate(std::max(_timeout, uint32_t{1}));
       });
 }
 
+ClusterMetricsFeature::Result ClusterMetricsFeature::readData(
+    uint64_t& version) {
+  auto vocbase = server().getFeature<SystemDatabaseFeature>().use();
+  aql::QueryString readText{
+      absl::StrCat("FOR m IN ", StaticStrings::MetricsCollection, " RETURN m")};
+  auto readQuery = aql::Query::create(
+      transaction::StandaloneContext::Create(*vocbase), readText, nullptr);
+  auto r = readQuery->executeSync();
+  if (r.fail()) {
+    return Result::Error;
+  }
+  auto metrics = r.data->slice().at(0);
+  version = metrics.get("version").getNumber<uint64_t>();
+  auto data = std::atomic_load_explicit(&_data, std::memory_order_acquire);
+  if ((data ? data->version : 0) < version) {
+    auto raw = metrics.get("data");
+    std::atomic_store_explicit(&_data, Data::fromVPack(version, raw),
```

Comment:
`Data::fromVPack(...)` may return `nullptr` in case of error. In this case `_data` will be replaced with `nullptr` which is bad.

Reply:
Maybe I should remove return nullptr?